### ISSUE TITLE
mf_priority: handle job updates

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-accounting#377]
As noted in #376, the mf_priority plugin should probably add a `job.update` callback so that it can handle updates to jobspec for pending jobs.

The `job.update` callback is issued after any `jobspec-update` event is processed, so the jobspec passed in the plugin args will have the updated values. An `updates` key is also passed in to the plugin args so this object could be scanned to see if any keys of interest were updated before reprocessing the jobspec, but that is probably optional for now.

The `job.update` callback is probably going to duplicate much of the code currently in the `job.new` callback, so possible a good solution here would be to split off the common code into a function that can be called by both `job.new` and `job.update`.

**Comment by grondo** [comment]
PR flux-framework/flux-core#5424 is ready to merge once the accounting priority plugin is capable of handling the queue update. I wonder if we'll want to prepare a WIP PR here, then merge the flux-core changes so that the implementation here can be tested...

**Comment by cmoussa1** [comment]
Sure, I can try to work on this this week and see if I can get something posted soon! Is there a way to for me to work in an interactive docker container with that specific flux-core branch? Not sure if I've ever done that before...

**Comment by grondo** [comment]
Oh, good question. What I would do is create the docker container then build and install that branch with something like:
```console
$ ./configure    --prefix=/usr  --sysconfdir=/etc  --with-systemdsystemunitdir=/etc/systemd/system --localstatedir=/var  --with-flux-security
$ sudo make -j 8 install
```

**Comment by cmoussa1** [comment]
Sorry for leaving you hanging on this - I definitely forgot to acknowledge your suggestion yesterday and didn't try this until just now - think I was able to get it working though.

(notes for myself for the future): I cloned your branch, launched an interactive Docker container, built and installed with the above commands, cloned the appropriate **flux-accounting** branch, built that, and was able to use the custom flux-core to do testing. 👍 thanks!!

**Comment by cmoussa1** [comment]
Just doing some super rudimentary testing this morning and wanted to report my findings so far.

This is specific to queue updates, but at least in my interactive Docker container, there was not a need to add a `job.update` callback to the priority plugin to be able to validate updated queues; I _think_ this is because as we mentioned in #378, after a job update, the job will re-enter `job.validate`, which _does_ check queue values passed in for a user/bank. The following set of tests I've been running attempt a workflow of submitting a pending job and updating its queue:

```shell
test_expect_success 'submit job for testing' '
	jobid1=$(flux python ${SUBMIT_AS} 5001 --urgency=0 --queue=bronze sleep 30)
'

test_expect_success 'update of queue of pending job works' '
	flux update $jobid1 queue=silver &&
	flux job eventlog $jobid1 > eventlog.out &&
	grep "attributes.system.queue=\"silver\"" eventlog.out
'
```

In this above workflow, a user `5001` is created with access to two (**bronze** and **silver**) of three available queues: **bronze**, **silver**, and **gold**. The job is submitted to the **bronze** queue and then updated to use the **silver** queue. So, the job re-enters `job.validate`, where the plugin will validate the specified queue here:

```c++
    bank_it->second.queue_factor = get_queue_info (queue, bank_it);

    if (bank_it->second.queue_factor == INVALID_QUEUE)
        return flux_jobtap_reject_job (p, args, "Queue not valid for user: %s",
                                       queue);
```

But, the user has access to the **silver queue**, the job gets validated with the new queue and re-enters `job.state.priority` with this new queue (I tested this by just printing out the unpacked value in the `job.state.priority` callback):

```
unpacked queue in job.state.priority: silver
```

Then, the user `5001` tries to update that same job to use the **gold** queue, which they should not have access to. So, the job will re-enter `job.validate` again, but this time, a `job-update` error is raised:

```
flux-update: ERROR: Queue not valid for user: gold
```

---

One point that I'm not sure about (probably warrants further discussion?) is that if the job re-enters `job.validate` and validation "fails" (because the user does not have access to the queue they're trying to update), the callback will call `flux_jobtap_reject_job ()`, but the job will not be rejected. In fact, its job update will not process, the queue will remain at the one it was at before the failed update, and it will stay pending:

```
flux-update: ERROR: Queue not valid for user: gold

1694709676.387470 submit userid=5001 urgency=0 flags=0 version=1
1694709676.399023 validate
1694709676.410215 depend
1694709676.410259 priority priority=0
1694709676.485914 jobspec-update attributes.system.queue="silver" attributes.system.constraints={}

       JOBID QUEUE    USER     NAME       ST NTASKS NNODES     TIME INFO
     ffPnQGw silver   5001     sleep       S      1      -        - 
```

Sorry if I am just completely missing something else here; I recognize that maybe the plugin will need a `job-update` callback for flux-accounting specific updates (like bank in the future?), but these were my findings with updating the queues.

**Comment by grondo** [comment]
> It does change when the queue changes and the job re-enters `job.validate`.

I think there's a problem here if the job queue update is eventually rejected by this or another `job.validate` callback. The queue factor has been updated to the new queue, but then the update is eventually denied, so the queue stays the same, but the queue factor has been changed. (Correct me if I'm wrong here)

I think the priority plugin may need the following changes:

 - In the `job.validate` callback, only _validate_ the job, do not make any changes that persist
 - Move the actual updates to a job as they relate to the priority to `job.new` and `job.update`, as theses are called for a valid job at submission and after an update respectively.

Does that make any sense?

**Comment by cmoussa1** [comment]
Ah, yes, I think I see what you mean. Sorry for not understanding you earlier. OK, I can see the plugin making those changes to start: remove assigning the associated queue priority to the `bank_info` struct associated with the job in `job.validate`: move it to `job.new` (and add it to a `job.update` callback).

I'll work on this!

**Comment by cmoussa1** [comment]
I think this can be closed now that #389 has been merged. A more specific issue (#391) has been opened to further expand supporting job updates for flux-accounting-specific attributes, and additional issues can be opened in the future if need be.
